contributor-docs/adrs/adr-005-box-sx.md

```diff
@@ -0,0 +1,62 @@
+# ADR 005: Use Box as a building block for all other components
+
+## Status
+
+Proposed
+
+## Context
+
+In Primer React and consuming applications, we use many different patterns for creating React components. Two common patterns are:
+
+1. Creating components with styled-components
+
+    ```tsx
+    const Avatar = styled.img.attrs<StyledAvatarProps>(props => ({
+      height: props.size,
+      width: props.size
+    }))<StyledAvatarProps>`
+      display: inline-block;
+      overflow: hidden;
+      line-height: ${get('lineHeights.condensedUltra')};
+      border-radius: ${props => getBorderRadius(props)};
+      ${sx}
+    `
+    ```
+
+    [Show full code example →](https://github.com/primer/react/pull/2019/files?diff=split&w=0)
+
+2.  Creating components with Box
+
+    ```tsx
+    const Avatar: React.FC<AvatarProps> = ({size = 20, alt = '', square = false, sx = {}, ...props}) => {
+      const styles = {
+        display: 'inline-block',
+        overflow: 'hidden',
+        lineHeight: 'condensedUltra',
+        borderRadius: getBorderRadius({size, square})
+      }
```

Comment:
Our goals as customers/consumers are:
1. Ease-of-use in terms of authorship (Be it specifying colors, styles, customization, etc)
2. Implementing something that is performant with little to no friction
3. Strongly typed guardrails, so it's easier to implement ("Do I preface this color with `colors.`? Do I pass a theme here? What does 'space.2' mean?")

I see some pros and cons to this approach:
**Pros**:
- This allows consumers to more easily define their styles without as much overhead
- It's coupled with the component, so there is less mental friction
- It still provides a solution that is responsive based on props (i.e. dynamic styles)

**Cons**:
- The biggest drawback I see is that these styles, because they are implemented inside of the component, would be re-computed every single render cycle. While this might be improved with memoization, I still think there's a good deal of perf impact with this approach
- I'm not a huge fan of each component needing to run `merge()` to get styles to work properly, not to mention the added dependency that each app would need to install
- I'm concerned about the mental friction of implementing everything as a `Box` and transmuting them into other components by way of `as`. I always thought of that as an escape hatch more than a common pattern.

I definitely think there are improvements to be made on the implementations of styles, specifically styled-components, but I'm also curious to see what styled-components' plans are now that React 18 is released with all of their improvements and DX hooks for these sorts of libraries.

Reply:
I  don't think the intention is to require a dependency on merge @dusave but just that it's the consumers job to merge local and props sx if necessary?

if `Box` is a `Memo` component, could we avoid some overhead by doing a deeper comparison on sx props

```
memo(function Box(props) {...}, function shouldComponentRender(prevProps, nextProps) {
   compare sx deepy, but maintain as object.is equality otherwise
})
```

which might avoid some/many performance differences from inline definitions of sx?

Reply:
> I'm concerned about the mental friction of implementing everything as a Box and transmuting them into other components by way of as. I always thought of that as an escape hatch more than a common pattern.

this, i think would be just something to extract to module scope and use from there?

Reply:
> Cons:

I'm glad we're talking about this. the sx implementation could use a lot of improvements!

> The biggest drawback I see is that these styles, because they are implemented inside of the component, would be re-computed every single render cycle. While this might be improved with memoization, I still think there's a good deal of perf impact with this approach

You're not wrong. We need work performance improvements to make sure the page doesn't slow down over time because of re-renders.

However, injecting styles is the biggest task in the whole process. Because Box still uses styled-components to inject styles in the end, the same styles are not injected twice. The exception to this is prop-based styles which need to be re-evaluated and the additional work required by styled-system on top of styled-components could be a bad fit for use cases like javascript-based animations.

> I'm not a huge fan of each component needing to run merge() to get styles to work properly, not to mention the added dependency that each app would need to install

Same. I have a few API ideas on how to simplify the API and abstract the merge away from the component.

For the merge dependency, right now it comes from `import {merge} from '@primer/react'`, but with API changes we should be able to remove it from the public API.

> I'm concerned about the mental friction of implementing everything as a Box and transmuting them into other components by way of as. I always thought of that as an escape hatch more than a common pattern.

Can you expand on the friction a bit more.

I think you only need to drop down to Box when you're building a new atomic component. `<Box as="img">` is equivalent to `styled.img`.

For most use cases, you should be able to start with a primer/react component, especially as we get more layout components like Stack. However, to give better html semantics, `as` prop would be the way to go: `<Stack as="section"/>`